cloud.ganttproject.colloboque/src/main/kotlin/cloud/ganttproject/colloboque/ColloboqueServer.kt

```diff
@@ -23,52 +23,134 @@ import biz.ganttproject.core.io.parseXmlProject
+    }
+  }
+
+  fun init(projectRefid: ProjectRefid, debugCreateProject: Boolean) {
+    try {
+      dataSourceFactory(projectRefid).let { ds ->
+        ds.connection.use {
+          it.createStatement().executeQuery("SELECT uid FROM Task").use { rs ->
+            while (rs.next()) {
+              println(rs.getString(1))
+            }
           }
-        }
-        if (debugCreateProject) {
-          DSL.using(it, SQLDialect.POSTGRES)
-            .configuration()
-            .deriveSettings { it.withRenderNameCase(RenderNameCase.LOWER) }
-            .dsl().let { dsl ->
+          if (debugCreateProject) {
+            DSL.using(it, SQLDialect.POSTGRES)
+              .configuration()
+              .deriveSettings { it.withRenderNameCase(RenderNameCase.LOWER) }
+              .dsl().let { dsl ->
 
-              loadProject("""
+                loadProject(
+                  """
 <?xml version="1.0" encoding="UTF-8"?>
 <project name="" company="" webLink="" view-date="2022-01-01" view-index="0" gantt-divider-location="374" resource-divider-location="322" version="3.0.2906" locale="en">
     <tasks empty-milestones="true">
         <task id="0" uid="qwerty" name="Task1" color="#99ccff" meeting="false" start="2022-02-10" duration="25" complete="85" expand="true"/>
     </tasks>
 </project>
-          """.trimIndent(), dsl)
-            }
+          """.trimIndent(), dsl
+                )
+              }
+          }
+          refidToBaseTxnId[projectRefid] = "abacaba"  // TODO: get from the database
         }
       }
+    } catch (e: Exception) {
+      throw ColloboqueServerException("Failed to init project $projectRefid", e)
     }
   }
+
+  private suspend fun processUpdatesLoop() {
+    while (true) {
+      val inputXlog = updateInputChannel.receive()
+      try {
+        val newBaseTxnId = commitTxnIfSucc(inputXlog.projectRefid, inputXlog.baseTxnId, inputXlog.transactions[0])
+          ?: continue
+        val response = ServerCommitResponse(
+          inputXlog.baseTxnId,
+          newBaseTxnId,
+          inputXlog.projectRefid,
+          SERVER_COMMIT_RESPONSE_TYPE
+        )
+        serverResponseChannel.send(Json.encodeToString(response))
+      } catch (e: Exception) {
+        LOG.error("Failed to commit\n {}", inputXlog, e)
+        val errorResponse = ServerCommitError(
+          inputXlog.baseTxnId,
+          inputXlog.projectRefid,
+          e.message.orEmpty(),
+          SERVER_COMMIT_ERROR_TYPE
+        )
+        serverResponseChannel.send(Json.encodeToString(errorResponse))
+      }
+    }
+  }
+
+  /** Performs transaction commit if `baseTxnId` corresponds to the value hold by the server. */
+  private fun commitTxnIfSucc(projectRefid: ProjectRefid, baseTxnId: String, transaction: XlogRecord): ProjectRefid? {
+    if (transaction.sqlStatements.isEmpty()) throw ColloboqueServerException("Empty transactions not allowed")
+    if (getBaseTxnId(projectRefid) != baseTxnId) return null
```

Comment:
Isn't it a normal case which can be ignored? A client could send multiple requests with the same update (either because few events happened or we do it periodically), the first one is processed, others can be just skipped

Reply:
Nope. What are possible client side strategies with respect to ? 
1. "Synchronized" request-response: a) generate UPDATE... > b) send Xlog > c) pause sending until commit response comes > d) update base txn id > e) send new updates, if any, with the new base txn id. If everything is ok, it sends a new update with a correct base txn id, but should it happen to be different, the client needs to know, because it has to resolve the issue and send a new update again
2. "Asynchronous" request-response: a) generate UPDATE... > b) send Xlog > c) generate UPDATE ... > d) send Xlog again > e) receive commit response to "b)". Here at the moment d) we still can't just ignore the Xlog, because the client needs to send it again